### PR TITLE
AEROGEAR-2415 Custom Runtime APB changes

### PR DESCRIPTION
**Pull request by ghost**
- [ ] The client id should be passed as a param to the bind
- [ ] The binding playbook will need to be created along with unbind
- [ ] The provision playbook should create a configmap which it uses during the bind to create a new configmap labelled with the client id
- [ ] Need to consider multiple clients if we take the approach of a master configmap and someone edits that configmap (is there a better approach)
- [ ] the deprovision playbook should delete the master configmap
- [ ] writing errors to the [/dev/termination-log](https://github.com/openshift/ansible-service-broker/issues/889)

### CODE REVIEW

File: roles/bind-crc-apb/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+---
+- name: Retrieve {{ _apb_provision_creds.SERVICE_NAME }} service configuration
+  shell: oc get configmaps/{{ crc_mobile_config_name }} -n {{ namespace }} -o jsonpath='{.data.config}'
+  register: svc_json_str_config
+
+- name: Validate CRC JSON config
+  set_fact: svc_json_config={{ svc_json_str_config.stdout | from_json }}
+  register: crc_json_config_valid
+  ignore_errors: True
```
Comment (maleck13) [inline roles/bind-crc-apb/tasks/main.yml:12]: should we be ignoring errors here?

File: roles/bind-crc-apb/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  k8s_v1_config_map:
+    name: '{{ CLIENT_ID | lower }}'
+    namespace: '{{ namespace }}'
+    annotations:
+      configType: '{{ crc_mobile_config_type }}'
+    labels:
+      name: '{{ _apb_provision_creds.SERVICE_NAME }}'
+      mobile: '{{ crc_mobile_svc_label }}'
+      serviceName: '{{ crc_mobile_svc_name }}'
+      mobileClientId: '{{ CLIENT_ID }}'
```
Comment (maleck13) [inline roles/bind-crc-apb/tasks/main.yml]: we should label this with the serviceInstanceID also https://github.com/aerogearcatalog/keycloak-apb/pull/62/files#diff-e9821aec58d768e751704303ab5a0875R143

File: roles/bind-crc-apb/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+- name: Creating CRC {{ CLIENT_ID }} configmap
+  k8s_v1_config_map:
+    name: '{{ CLIENT_ID | lower }}'
+    namespace: '{{ namespace }}'
+    annotations:
+      configType: '{{ crc_mobile_config_type }}'
+    labels:
+      name: '{{ _apb_provision_creds.SERVICE_NAME }}'
+      mobile: '{{ crc_mobile_svc_label }}'
```
Comment (maleck13) [inline roles/bind-crc-apb/tasks/main.yml]: not sure on this label, it should be mobile:enabled https://github.com/aerogearcatalog/keycloak-apb/pull/62/files#diff-d03fc7e3ea5b47a1f847b645bdf93da8R7

File: roles/provision-crc-apb/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
+---
+- name: Validate CRC JSON config
+  set_fact: svc_json_config={{ SERVICE_CONFIG | from_json }}
+  register: crc_json_config_valid
+  ignore_errors: True
+
+- fail:
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: I think in this case we should use last_operation so the user is informed of the reason for the failure.

```
asb_last_operation:
    description: "config was not valid json. Could not provision"  
```

We can use last_operation throughout the apb to inform the user of what is happening during a provision

File: roles/provision-crc-apb/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  when: crc_json_config_valid|failed
+
+- name: Creating CRC {{ crc_mobile_config_name }} secrets
+  k8s_v1_secret:
+    name: '{{ crc_mobile_config_name }}'
+    namespace: '{{ namespace }}'
+    labels:
+      name: '{{ SERVICE_NAME }}'
+      mobile: '{{ crc_mobile_svc_label }}'
+      serviceName: '{{ crc_mobile_svc_name }}'
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: same as above should be ```mobile:enabled``` and also have serviceInstanceID label

File: roles/provision-crc-apb/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+- name: Creating CRC {{ crc_mobile_config_name }} configmap
+  k8s_v1_config_map:
+    name: '{{ crc_mobile_config_name }}'
+    namespace: '{{ namespace }}'
+    annotations:
+      configType: '{{ crc_mobile_config_type }}'
+    labels:
+      name: '{{ SERVICE_NAME }}'
+      mobile: '{{ crc_mobile_svc_label }}'
+      serviceName: '{{ crc_mobile_svc_name }}'
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: same here

File: roles/provision-crc-apb/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    data:
+      type: '{{ crc_mobile_svc_type }}'
+      name: '{{ SERVICE_NAME }}'
+      uri: '{{ SERVICE_URI }}'
+      config: '{{ svc_json_config | to_nice_json }}'
+
+- name: Get the name of the service instance
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: we don't need to do this as, we are doing work to remove the need for this labeling as part of this sprint

File: roles/provision-crc-apb/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  shell: >
+    oc get serviceinstance
+    -o jsonpath='{.items[?(@.spec.externalID=="{{ _apb_service_instance_id }}")].metadata.name}'
+    --namespace={{ namespace }}
+  when: _apb_service_instance_id is defined
+  register: service_instance_name
+
+- name: Label the service instance with the service name
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: remove

File: roles/provision-crc-apb/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  shell: >
+    oc label serviceinstance {{ service_instance_name.stdout }}
+    serviceName={{ crc_mobile_svc_name }}
+    --namespace={{ namespace }}
+  when: _apb_service_instance_id is defined
+
+- name: Label the service instance with mobile {{ crc_mobile_svc_label }}
```
Comment (maleck13) [inline roles/provision-crc-apb/tasks/main.yml]: remove

**Review by maleck13 — CHANGES_REQUESTED** [review]
This is a pretty good start but some changes are needed. As discussed at sprint planning we are removing the need to label the service instance.

Also could you add some steps for verification

File: roles/bind-crc-apb/tasks/main.yml (hunk 1 above)
Comment (ghost) [reply roles/bind-crc-apb/tasks/main.yml:12]: This one is ignored as following FAIL will fail when JSON is not valid

File: roles/bind-crc-apb/tasks/main.yml (hunk 2 above)
Comment (ghost) [reply roles/bind-crc-apb/tasks/main.yml]: labeled with serviceInstanceID

File: roles/bind-crc-apb/tasks/main.yml (hunk 3 above)
Comment (ghost) [reply roles/bind-crc-apb/tasks/main.yml]: _crc_mobile_svc_label_ variable is moved to _vars/_ fodler and is crc_mobile_svc_label: 'enabled'. Name can be confusing, suggestions?

Comment (maleck13) [reply roles/bind-crc-apb/tasks/main.yml]: Ahh ok. No thats fine thanks for explaining

File: roles/provision-crc-apb/tasks/main.yml (hunk 4 above)
Comment (ghost) [reply roles/provision-crc-apb/tasks/main.yml]: We can use it instead of fail? Or we need to use it before fail?

**Comment by ghost** [comment]
**VALIDATION:**

- [x]     Provision Custom Runtime Connector APB
- [x]     Wait until service has READY state
- [x]     Check if master configmap is created
- [x]     Check if master secret is created
- [x]     Bind service and specify CLIENT_ID
- [x]     Check if CLIENT configmap is created
- [x]     Unbind service
- [x]     Check if CLIENT configmap is deleted
- [x]     Delete service and wait until deprovisioned
- [x]     Check if master configmap is deleted
- [x]     Check if master secret is deleted

File: roles/provision-crc-apb/tasks/main.yml (hunk 4 above)
Comment (maleck13) [reply roles/provision-crc-apb/tasks/main.yml]: ok so I think a better solution (we should use last operation for each step happening) for the error case is to write the error to ```/dev/termination-log``` this is read by k8s and set as the termination status for the pod. This is checked by the broker and displayed if present

File: roles/bind-crc-apb/tasks/main.yml (hunk 1 above)
Comment (odra) [reply roles/bind-crc-apb/tasks/main.yml:12]: The `failed_when` directive would work here as well

**Comment by maleck13** [comment]
👀

**Comment by maleck13** [comment]
I think we need to change the name of the configmaps created to be ```serviceName-master``` ```serviceName-test``` to add some kind of namespacing and avoid clashes when more than one CRC is able to be used (which we are doing in this sprint)

I also would like to to see us try writing errors to the /dev/termination-log so that make their way back to the user

File: roles/bind-crc-apb/tasks/main.yml (hunk 1 above)
Comment (ghost) [reply roles/bind-crc-apb/tasks/main.yml:12]: @odra this is intentional, we want custom messages to be used in next FAIL task

**Comment by maleck13** [comment]
Could we change the message shown on bind from
``` Client's Id To Copy Service JSON Configuration```
to just
```Mobile Client Id ```

**Comment by maleck13** [comment]
This is ok to be merged once the binding text is changed

**Review by maleck13 — APPROVED** [review]

**Pull request merged**